### ISSUE TITLE
Textless_nlp / gslm: Which CPC embedding should I use in order to use the pretrained k-means model?

### ISSUE THREAD

**Issue by jaeyeonkim99** [facebookresearch/fairseq#3988]
#### What is your question?

Hello.I want to do the job somewhat similar to the task in the paper "Generative Spoken Language Modeling from Raw Audio". I want to get CPC embedding for the raw audio, quantize it with k-means, and train the language model with the quantized embeddings. In this process, I tried to use the pretrained CPC model(Modified CPC) and K-means model from https://github.com/pytorch/fairseq/tree/main/examples/textless_nlp/gslm/speech2unit 

However, I  cannot find the specification about which layer to use in the CPC model(in the paper, it is only mentioned that intermediate layer is used). I tried to get features by giving layer=None option to CpcFeatureReader(which seams to be the default value), but CPC embeddings generated doesn't match the downloaded K-means model.(Below picture is the t-SNE result: blue points represent the k-means centroids, red points represent the CPC embeddings(option: layer = None))

https://github.com/jaeyeonkim99/Images/blob/main/CPC/CPC_embedding_with_clusters.png

How can I get the appropriate CPC embedding corresponding to pretrained K-means? Which layer should I use, and is there any additional options that I should give?

Thank you.

#### Code

CPCFeatureReader is from https://github.com/pytorch/fairseq/tree/main/examples/textless_nlp/gslm/speech2unit/pretrained

```python
self.phone_encoder = CpcFeatureReader(checkpoint_path=self.phone_pretrain_path, layer = None)
self.kmeans_model = joblib.load(open(self.kmeans_path, "rb"))
self.kmeans_model.verbose = False
   
#get CPC feature
cpc_embedding = self.phone_encoder.get_feats(audio)
cpc_embedding = cpc_embedding.cpu().numpy()

#Quantize Using Kmeans
phone_embedding = self.kmeans_model.predict(cpc_embedding)
   
```

#### What's your environment?

 - fairseq Version: 0.10.2	
 - PyTorch Version: 1.7.1+cu110
 - OS (e.g., Linux): Linux(Ubuntu 16.04)
 - How you installed fairseq (`pip`, source): source
 - Python version: 3.8.8
 - CUDA version: 11.0
 - GPU models and configuration: GTX 1080

**Comment by cdaube** [comment]
Stumbled over the same thing.

The arxiv GSLM paper  mentions that they used "an intermediate layer of the predictor, which provides a 256-dimensional embedding" (p. 6). However, as far as I can tell, there is no layer (neither in the convolutional encoder nor in the gAR on top of it, where the "layer" argument defines only the number of layers in the latter if I understand correctly) that has 256 dimensions -- they all seem to be 512-D. This dimensionality however seems to correspond to the kmeans centroids provided in the kmeans model, which is 512-D, so 256 as in the paper seems definitely wrong.

I did a check where I encoded a wav file containing speech into the pretrained quantised CPC space (and also w2v2, hubert and logmel using the layer specifications as detailed in the arxiv paper, i.e. n.a. for logmel, 14 for w2v2 and 6 for hubert). I then computed entropy across time as a rough metric to quantify whether the output is well-mixed across the classes, and found that for w2v2, hubert and logmel, these entropies across time are highly similar across different quantisations (50, 100, 200), and result in well-mixed predictions. 

For CPC, I get something like that when I choose **layer = 2** , and the entropy is lower (i.e. the kmeans pretty much always assigns the same class) when I in- or decrease the gAR layer argument, making me think that 4 (which corresponds to the default, i.e. what we get with layer = None) is most definitely not what was used for the models downstream (and of course values higher than what is available from the pretrained weights will just give nonsense).

Here is a plot that shows what I'm describing above (layer = 0 is the output straight from the convolutional encoder):
![Screenshot from 2022-04-13 18-48-33](https://user-images.githubusercontent.com/32395102/163240206-c985eca7-bf9f-43c7-a4b3-ac6a90891514.png)

For now, my best shot would thus be using layer = 2 for CPC.
But would still be cool to know from the authors whether this is correct :)

**Comment by hikushalhere** [comment]
About the choice of layer in gAR module, the second layer is indeed the layer we had used in GSLM.  see that unlike wav2vec 2.0 and HuBERT, we haven't specified the choice of layer for CPC in the GSLM paper. It makes sense to explicitly note the layer for CPC too. Thanks @jaeyeonkim99 and @cdaube for the feedback.
